File: examples/bridges/python_peer/rim_router_peer.py
"""UDS bridge for a Python peer talking to the C++ router.

A ``RouterPeer`` owns one ``AF_UNIX`` ``SOCK_DGRAM`` socket bound at the
peer's UDS path declared in the topology profile. Every frame is one
64 B datagram (no length prefix, no stream framing — the router uses
``SOCK_DGRAM`` for natural message boundaries).

How the router identifies us
----------------------------
The C++ router resolves the source peer from the *sender's bound socket
path*, not from the ``source`` byte in the frame. See
``peer_id_from_recv`` in ``ipc/src/router/datagram_peer_resolver.hpp``:
the router calls ``recvfrom`` and matches the returned ``sun_path``
against ``PeerEntry::local.u.uds_path``. So we **must** ``bind`` to the
peer's declared path before sending, or the router will count our
datagrams as ``recv_unknown_source`` (the Phase D4 fault counter) and
drop them.

References
----------
- ipc/src/router/link.hpp                       C++ peer-side client mirror
- ipc/src/router/datagram_peer_resolver.hpp     source-id resolution
- ipc/src/ipc/datagram.hpp                      UDS socket layer
- docs/adr/0008-router-frame-v2.md              wire format
- examples/bridges/README.md                    bridge boundary policy
"""

from __future__ import annotations

import os
import socket
from pathlib import Path
from typing import Optional

from rim_router_frame import ROUTER_FRAME_SIZE, RouterFrame


class RouterPeer:
    """Bound UDS endpoint for a Python peer.

    Parameters
    ----------
    router_path:
        UDS path the router server is listening on (``[router].listen`` in
        the profile TOML, e.g. ``/tmp/rim_router.sock``).
    peer_path:
        UDS path declared for *this* peer in ``[[peers]]``. The router will
        ``sendto`` this path when fanning out frames; we must bind here.
    peer_id:
        Optional. Stored for self-stamping the ``source`` byte on
        publishes — purely informational because the router rewrites
        ``source`` on forward.
    recv_timeout:
        Optional. Seconds for blocking ``recv()``; ``None`` = blocking,
        ``0.0`` = non-blocking, anything > 0 maps to ``SO_RCVTIMEO``.

    Lifetime
    --------
    The bound socket path is unlinked on close. Use as a context manager
    or call ``close()`` explicitly to avoid leaving stale sockets that
    would block re-bind on restart (same hazard the C++ router solves
    with the ``unlink`` call at the top of ``Uds::bind``).
    """

    def __init__(
        self,
        router_path: str,
        peer_path: str,
        peer_id: int = 0,
        recv_timeout: Optional[float] = None,
    ) -> None:
        self.router_path = router_path
        self.peer_path = peer_path
        self.peer_id = peer_id
        self._sock: Optional[socket.socket] = None
        self._closed = False

        self._open(recv_timeout)

    def _open(self, recv_timeout: Optional[float]) -> None:
        Path(self.peer_path).parent.mkdir(parents=True, exist_ok=True)
        try:
            os.unlink(self.peer_path)
        except FileNotFoundError:
            pass
        except OSError as exc:
            raise RuntimeError(
                f"could not unlink stale peer socket {self.peer_path!r}: {exc}"
            ) from exc

        self._sock = socket.socket(socket.AF_UNIX, socket.SOCK_DGRAM)
        try:
            self._sock.bind(self.peer_path)
        except OSError as exc:
            self._sock.close()
            self._sock = None
            raise RuntimeError(
                f"bind {self.peer_path!r} failed: {exc}"
            ) from exc

        if recv_timeout is not None:
            self._sock.settimeout(recv_timeout)

    def set_recv_timeout(self, seconds: Optional[float]) -> None:
        """Set blocking recv timeout. ``None`` = blocking, ``0`` = non-blocking."""
        assert self._sock is not None, "socket closed"
        self._sock.settimeout(seconds)

    def fileno(self) -> int:
        assert self._sock is not None, "socket closed"
        return self._sock.fileno()

    def send_frame(self, frame: RouterFrame) -> None:
        """Send a fully-constructed frame to the router as one datagram."""
        assert self._sock is not None, "socket closed"
        data = frame.to_bytes()
        if len(data) != ROUTER_FRAME_SIZE:
            raise RuntimeError(
                f"frame is not exactly {ROUTER_FRAME_SIZE} bytes (got {len(data)})"
            )
        self._sock.sendto(data, self.router_path)

    def publish(
        self,
        payload: bytes | str = b"",
        *,
        topic_id: int = 0,
        flags: int = 0,
        seq: int = 0,
        timestamp_ns: int = 0,
        priority: int = 0,
    ) -> RouterFrame:
        """Convenience: build a frame and send it. Returns the frame sent."""
        if priority < 0 or priority > 7:
            raise ValueError("priority must fit in 3 bits (0..7)")
        full_flags = flags & 0x8F
        full_flags |= (priority & 0x07) << 4

        frame = RouterFrame.make(
            source=self.peer_id,
            flags=full_flags,
            topic_id=topic_id,
            seq=seq,
            timestamp_ns=timestamp_ns,
            payload=payload,
        )
        self.send_frame(frame)
        return frame

    def recv_frame(self) -> Optional[RouterFrame]:
        """Block (or wait until timeout) for one datagram. Returns ``None`` on timeout.

        Truncated datagrams (< 64 B) are dropped silently to match the
        router's Phase D4 ``recv_truncated`` policy on the server side.
        Downstream callers see ``None`` as "no valid frame this tick".
        """
        assert self._sock is not None, "socket closed"
        try:
            data, _ = self._sock.recvfrom(ROUTER_FRAME_SIZE)
        except (socket.timeout, BlockingIOError):
            return None
        if len(data) < ROUTER_FRAME_SIZE:
            return None
        return RouterFrame.from_bytes(data[:ROUTER_FRAME_SIZE])

    def close(self) -> None:
        if self._closed:
            return
        self._closed = True
        if self._sock is not None:
            try:
                self._sock.close()
            finally:
                self._sock = None
        try:
            os.unlink(self.peer_path)
        except FileNotFoundError:
            pass

    def __enter__(self) -> "RouterPeer":
        return self

    def __exit__(self, exc_type, exc, tb) -> None:
        self.close()

    def __del__(self) -> None:
        try:
            self.close()
        except Exception:
            pass


def router_now_ns() -> int:
    """Match the router's ``CLOCK_MONOTONIC_RAW`` source (ADR 0010).

    Python 3.7+ provides ``time.clock_gettime_ns(time.CLOCK_MONOTONIC_RAW)``
    on Linux; falls back to ``time.monotonic_ns()`` on platforms that
    don't expose ``CLOCK_MONOTONIC_RAW`` (BSDs, macOS). Cross-host
    correlation is explicitly delegated per ADR 0010 — this helper is
    only useful for single-host comparable timestamps.
    """
    import time

    if hasattr(time, "CLOCK_MONOTONIC_RAW"):
        return time.clock_gettime_ns(time.CLOCK_MONOTONIC_RAW)
    return time.monotonic_ns()

File: examples/bridges/node_gateway/rim_router_peer.js
// UDP bridge for a Node peer talking to the C++ router.
//
// A `RouterPeer` owns one `dgram.Socket` ('udp4') bound at the peer's UDP
// host:port declared in the topology profile (e.g. `hil.toml` peer 8 is at
// `127.0.0.1:19108`). Every frame is one 64 B datagram (no length prefix,
// no stream framing — the router uses `SOCK_DGRAM` for natural message
// boundaries on UDP just as it does on UDS).
//
// How the router identifies us
// ----------------------------
// The C++ router resolves the source peer from the sender's UDP source
// host:port, not from the `source` byte in the frame. See
// `peer_id_from_recv<Udp>` in `ipc/src/router/datagram_peer_resolver.hpp`:
// the router calls `recvfrom` and matches the returned `sockaddr_in`
// host + port against `PeerEntry::local.u.udp`. So we MUST `bind` to the
// peer's declared host:port before sending, or the router will count our
// datagrams as `recv_unknown_source` (the Phase D4 fault counter) and
// drop them.
//
// Why not UDS?
// ------------
// Node's stdlib `dgram` module supports only UDP. `AF_UNIX SOCK_DGRAM`
// (the transport the router uses for UDS profiles like `x86_dev.toml`)
// has no Node stdlib equivalent — Node's `net` covers `AF_UNIX` but only
// as `SOCK_STREAM`. UDS-DGRAM access in Node requires either the
// `unix-dgram` npm package or a small N-API binding. See the
// node_gateway/README.md "Transport reachability" section for the full
// matrix and the opt-in path.
//
// References:
//   - ipc/src/router/link.hpp                       C++ peer-side client mirror
//   - ipc/src/router/datagram_peer_resolver.hpp     source-id resolution
//   - ipc/src/ipc/datagram.hpp                      UDP socket layer
//   - docs/adr/0008-router-frame-v2.md              wire format
//   - examples/bridges/python_peer/rim_router_peer.py   peer port (Python)

import dgram from 'node:dgram';
import { EventEmitter } from 'node:events';

import {
    ROUTER_FRAME_SIZE,
    RouterFrame,
    makeFrame,
} from './rim_router_frame.js';

export class RouterPeer extends EventEmitter {
    // Construction does not bind the socket; callers must `await open()`
    // first. This mirrors the synchronous-bind discipline of the Python
    // peer but adapted to Node's async UDP socket lifecycle.
    constructor({ routerHost, routerPort, peerHost, peerPort, peerId = 0 } = {}) {
        super();
        if (!routerHost || !routerPort || !peerHost || !peerPort) {
            throw new Error(
                'RouterPeer requires routerHost, routerPort, peerHost, peerPort'
            );
        }
        this.routerHost = routerHost;
        this.routerPort = Number(routerPort);
        this.peerHost = peerHost;
        this.peerPort = Number(peerPort);
        this.peerId = Number(peerId) & 0xFF;
        this.sock = null;
        this.closed = false;
    }

    async open() {
        if (this.sock) {
            throw new Error('RouterPeer.open: already open');
        }
        const sock = dgram.createSocket({ type: 'udp4', reuseAddr: false });
        this.sock = sock;

        sock.on('message', (msg, rinfo) => {
            // Truncated datagrams (< 64 B) are dropped silently to match
            // the router's Phase D4 `recv_truncated` policy on the server
            // side. Downstream callers see no 'frame' event for those.
            if (msg.length < ROUTER_FRAME_SIZE) {
                this.emit('truncated', msg.length, rinfo);
                return;
            }
            try {
                const frame = RouterFrame.fromBytes(msg);
                this.emit('frame', frame, rinfo);
            } catch (err) {
                this.emit('error', err);
            }
        });

        sock.on('error', (err) => this.emit('error', err));

        await new Promise((resolve, reject) => {
            const onError = (err) => {
                sock.removeListener('listening', onListening);
                reject(new Error(
                    `bind ${this.peerHost}:${this.peerPort} failed: ${err.message}`
                ));
            };
            const onListening = () => {
                sock.removeListener('error', onError);
                resolve();
            };
            sock.once('error', onError);
            sock.once('listening', onListening);
            sock.bind({ address: this.peerHost, port: this.peerPort, exclusive: true });
        });
    }

    sendFrame(frame) {
        if (!this.sock) {
            throw new Error('RouterPeer.sendFrame: socket not open');
        }
        const data = frame.toBytes();
        if (data.length !== ROUTER_FRAME_SIZE) {
            throw new Error(
                `frame is not exactly ${ROUTER_FRAME_SIZE} bytes (got ${data.length})`
            );
        }
        return new Promise((resolve, reject) => {
            this.sock.send(data, 0, data.length, this.routerPort, this.routerHost, (err) => {
                if (err) reject(err); else resolve();
            });
        });
    }

    publish({
        payload = '',
        topic_id = 0,
        flags = 0,
        seq = 0,
        timestamp_ns = 0n,
        priority = 0,
    } = {}) {
        if (priority < 0 || priority > 7) {
            throw new RangeError('priority must fit in 3 bits (0..7)');
        }
        let fullFlags = flags & 0x8F;
        fullFlags |= (priority & 0x07) << 4;

        const frame = makeFrame({
            source: this.peerId,
            flags: fullFlags,
            topic_id,
            seq,
            timestamp_ns,
            payload,
        });
        return this.sendFrame(frame).then(() => frame);
    }

    async close() {
        if (this.closed) return;
        this.closed = true;
        if (this.sock) {
            const sock = this.sock;
            this.sock = null;
            await new Promise((resolve) => sock.close(resolve));
        }
    }
}

// Match the router's `CLOCK_MONOTONIC_RAW` timestamp source (ADR 0010) as
// closely as Node stdlib allows. Node's `process.hrtime.bigint()` returns
// nanoseconds since an arbitrary boot-anchored origin (CLOCK_MONOTONIC on
// Linux), which is NOT identical to CLOCK_MONOTONIC_RAW — MONOTONIC is
// subject to NTP slew, RAW is not. For dashboard / debug single-host use
// the difference is negligible (< 1 ppm typically). Cross-host correlation
// is explicitly delegated per ADR 0010; this helper is only useful for
// single-host comparable timestamps. The router rewrites `timestamp_ns`
// on forward anyway (see `run_forward_loop` in `router_server.cpp`), so
// publish-time stamps are mostly informational.
export function routerNowNs() {
    return process.hrtime.bigint();
}
